keras_nlp/models/causal_lm.py

```diff
@@ -387,3 +390,106 @@ def postprocess(x):
             outputs = [postprocess(x) for x in outputs]
 
         return self._normalize_generate_outputs(outputs, input_is_scalar)
+
+    """Wires LLM decoding caches into a backbone.
+    
+    Returns a new functional backbone with the same graph layout as the
+    original backbone, but with caches wired into TransformerDecoder blocks.
+    """
+    @staticmethod
+    def _rewire_backbone_with_cache(backbone, cache_shape, cache_dtype):
+
+        # Define new inputs for caches.
+        cache_update_index_input = keras.Input(
+            shape=(), dtype="int32", name="cache_update_index"
+        )
+        # cache_update_index_input is always a scalar. We must force the
+        # shape to scalar because keras.Input assumes a batch dim.
+        cache_update_index_input.shape = ()
+        # Input for a combined cache for all TransformerDecoder layers
+        cache_input = keras.Input(
+            shape=cache_shape, dtype=cache_dtype, name="cache"
+        )
+
+        # Split the cache on the num_layers axis=1 (number of transformer blocks).
+        caches = ops.unstack(cache_input, cache_input.shape[1], axis=1)
+        decoder_block_idx = 0
+        next_caches = []
+
+        def rewire_positionembedding(layer, *args, **kwargs):
+            # wire in a new input: cache_update_index_input by calling the layer on this input
+            nonlocal cache_update_index_input
+            return layer(*args, start_index=cache_update_index_input, **kwargs)
+
+        def rewire_transformerdecoder(layer, *args, **kwargs):
+            # wire in caches, next_caches by calling the layer on these inputs
+            nonlocal caches, next_caches, decoder_block_idx
+            # no mask when decoding with cache
+            kwargs.pop("decoder_padding_mask")
+            output, next_cache = layer(
+                *args,
+                self_attention_cache=caches[decoder_block_idx],
+                self_attention_cache_update_index=cache_update_index_input,
+                **kwargs,
+            )
+            decoder_block_idx += 1
+            next_caches.append(next_cache)
+            return output
+
+        def rewire_fn(layer, *args, **kwargs):
+            if isinstance(layer, PositionEmbedding):
+                return rewire_positionembedding(layer, *args, **kwargs)
+            elif isinstance(layer, TransformerDecoder):
```

Comment:
We can test for being in the set of backbone.transformer_blocks rather than a specific type. This can be solved with a convention of what "backbone" should contain (which makes sense - not any backbone works for cached text generation).